evm/src/cpu/kernel/asm/bignum/modmul.asm

```diff
@@ -0,0 +1,192 @@
+    %eq_const(@BIGNUM_LIMB_BASE)
+    // stack: val+1==limb_base, val+1, cur_loc, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    DUP1
+    // stack: val+1==limb_base, val+1==limb_base, val+1, cur_loc, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    ISZERO
+    // stack: val+1!=limb_base, val+1==limb_base, val+1, cur_loc, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    SWAP1
+    SWAP2
+    // stack: val+1, val+1!=limb_base, val+1==limb_base, cur_loc, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    MUL
+    // stack: to_write=(val+1)*(val+1!=limb_base), continue=val+1==limb_base, cur_loc, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    DUP3
+    // stack: cur_loc, to_write, continue, cur_loc, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    %mstore_kernel_general
+    // stack: continue, cur_loc, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    SWAP1
+    // stack: cur_loc, continue, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    %increment
+    // stack: cur_loc + 1, continue, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    DUP1
+    // stack: cur_loc + 1, cur_loc + 1, continue, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    DUP8
+    // stack: scratch_3, cur_loc + 1, cur_loc + 1, continue, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    EQ
+    // stack: cur_loc + 1 == scratch_3, cur_loc + 1, continue, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    ISZERO
+    // stack: cur_loc + 1 != scratch_3, cur_loc + 1, continue, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    SWAP1
+    SWAP2
+    // stack: continue, cur_loc + 1 != scratch_3, cur_loc + 1, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    MUL
+    // stack: new_continue=continue*(cur_loc + 1 != scratch_3), cur_loc + 1, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    %jumpi(increment_loop)
+    // stack: cur_loc + 1, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    POP
+no_carry:
+    // stack: length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+
+    // Calculate a * b.
+
+    // Store a * b in scratch_3.
+    PUSH modmul_return_3
+    // stack: modmul_return_3, length, a_start_loc, b_start_loc, scratch_2, scratch_3, retdest
+    %stack (return, len, a, b, s2, s3) -> (len, a, b, s3, return, len, s2, s3)
+    // stack: length, a_start_loc, b_start_loc, scratch_3, modmul_return_3, length, scratch_2, scratch_3, retdest
+    %jump(mul_bignum)
+modmul_return_3:
+    // stack: length, scratch_2, scratch_3, retdest
+
+    // Check that x + k * m = a * b.
+    // Walk through scratch_2 and scratch_3, checking that they are equal.
+    // stack: n=length, i=scratch_2, j=scratch_3, retdest
+modmul_check_loop:
```

Comment:
One more thing to check: `x < m`. Without it, yes, `x % m == (a * b) % m`, but that doesn't imply that indeed `x == a * b`. (Unless the weaker condition is fine, in which case it should go in a comment ☺️).